pol-core/bscript/berror.cpp

```diff
@@ -127,5 +133,123 @@ BObjectImp* BError::array_assign( BObjectImp* /*idx*/, BObjectImp* /*target*/, b
 {
   return this;
 }
+
+BObjectImp* BError::call_method_id( const int id, Executor& ex, bool )
+{
+  switch ( id )
+  {
+  case MTH_STACKTRACE:
+  {
+    bool returnAsObject = false;
+
+    if ( ex.numParams() > 0 && !ex.getParam( 0, returnAsObject ) )
+    {
+      return new BError( "Invalid parameter type" );
+    }
+
+    if ( returnAsObject )
+    {
+      if ( script_->read_dbg_file() != 0 )
+      {
+        return UninitObject::create();
+      }
+
+      std::unique_ptr<ObjArray> arr( new ObjArray );
+
+      for ( const auto& PC : stack_ )
+      {
+        auto filename = script_->dbg_filenames[script_->dbg_filenum[PC]];
+        auto line = script_->dbg_linenum[PC];
+        auto dbgFunction =
+            std::find_if( script_->dbg_functions.begin(), script_->dbg_functions.end(),
+                          [&]( auto& i ) { return i.firstPC <= PC && PC <= i.lastPC; } );
+
+        std::string functionName =
+            dbgFunction != script_->dbg_functions.end() ? dbgFunction->name : "<program>";
+
+        std::unique_ptr<BStruct> entry( new BStruct );
+        entry->addMember( "file", new String( filename ) );
+        entry->addMember( "line", new BLong( line ) );
+        entry->addMember( "name", new String( functionName ) );
+        arr->addElement( entry.release() );
+      }
+
+      return arr.release();
+    }
+
+    auto errortext = FindMember( "errortext" );
+
+    std::string message = errortext == nullptr ? "Error" : errortext->getStringRep();
+
+    if ( script_->read_dbg_file() == 0 )
+    {
+      for ( const auto& PC : stack_ )
+      {
+        auto filename = script_->dbg_filenames[script_->dbg_filenum[PC]];
+        auto line = script_->dbg_linenum[PC];
+        auto dbgFunction =
+            std::find_if( script_->dbg_functions.begin(), script_->dbg_functions.end(),
+                          [&]( auto& i ) { return i.firstPC <= PC && PC <= i.lastPC; } );
+
+        std::string functionName =
+            dbgFunction != script_->dbg_functions.end() ? dbgFunction->name : "<program>";
+
+        message.append( fmt::format( "\n    at {} ({}:{})", functionName, filename, line,
+                                     script_->dbg_filenames[script_->dbg_filenum[PC]],
+                                     script_->dbg_filenames[1] ) );
+      }
+    }
+    else
+    {
+      for ( const auto& pc : stack_ )
+      {
+        message.append( fmt::format( "\n    at {}+{}", Clib::scripts_thread_script, pc ) );
+      }
+    }
+
+    return new String( message );
+  }
+  default:
+    return base::call_method_id( id, ex, false );
+  }
+}
+
+void BError::attach_stack( Bscript::Executor* exec )
+{
+  if ( !exec )
+  {
+    return;
+  }
+
+  EScriptProgram* prog = const_cast<EScriptProgram*>( exec->prog() );
+  script_.set( prog );
+
+  if ( !stack_.empty() )
+  {
+    return;
+  }
+  std::vector<ReturnContext> callStack = exec->ControlStack;
```

Comment:
Since BError is one of the most created types I fear a bit the performance impact. Maybe have a setting for it?
In general this method could have some performance improvements. Don't copy ControlStack and instead reserve the size+1 add current PC and then iterate backwards over the ControlStack and add the pc.